src/runtime_src/core/tools/common/SubCmdJSON.cpp

```diff
@@ -0,0 +1,221 @@
+  allOptions.add_options()
+     ("help,h", boost::program_options::bool_switch(&help), "Help to use this sub-command")
+  ;
+
+  allOptions.add(hiddenOptions);
+
+  po::parsed_options parsed = po::command_line_parser(_options).
+    options(allOptions).            // Global options
+    positional(positionalCommand).  // Our commands
+    allow_unregistered().           // Allow for unregistered options (needed for sub options)
+    run();                          // Parse the options
+
+  po::variables_map vm;
+
+  try {
+    po::store(parsed, vm);
+    po::notify(vm); // Can throw
+  } catch (po::error& e) {
+    std::cerr << "ERROR: " << e.what() << std::endl << std::endl;
+    printHelp(commonOptions, hiddenOptions, true);
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+
+  // Check to see if no command was found
+  if ((vm.count("subCmd") == 0)) {
+    printHelp(commonOptions, hiddenOptions, true);
+    return;
+  }
+
+  // -- Now process the subcommand --------------------------------------------
+  std::string sCommand = vm["subCmd"].as<std::string>();
+
+  for ( auto &jsonCmd : subCmdOptions ) {
+    if (sCommand.compare(jsonCmd.option) == 0) {
+      std::vector<std::string> opts = po::collect_unrecognized(parsed.options, po::include_positional);
+      opts.erase(opts.begin());
+
+      if(help == true) {
+        opts.push_back("--help");
+      }
+
+      std::string finalCmd = jsonCmd.application + " " + jsonCmd.defaultArgs;
+      for (auto &opt : opts) {
+        finalCmd += " ";
+	finalCmd += opt;
+      }
+
+      std::cout << "\nInvoking application : " << jsonCmd.application << std::endl;
+      std::cout << "\ncommand : " << finalCmd << "\n\n";
+
+      int status = system(finalCmd.c_str());
+      if(status)
+        std::cout << "\nERROR: Failed to run the command - " << finalCmd << std::endl;
+
+      return;
+    }
+  }
+
+  std::cout << "\nERROR: Missing valid program operation. No action taken.\n\n";
+  printHelp(commonOptions, hiddenOptions, true);
+  throw xrt_core::error(std::errc::operation_canceled);
+}
+
+// ----- H E L P E R   F U N C T I O N S -------------------------------------------
+static void collectJsonPaths(std::vector<std::string> &pathVec, std::string env)
+{
+    char del = ':';
+    int start = 0;
+    int end = env.find(del);
+    while (end != -1) {
+        pathVec.emplace_back(env.substr(start, end - start));
+        start = end + 1;
+        end = env.find(del, start);
+    }
+    pathVec.emplace_back(env.substr(start, env.length() - start));
+}
+
+static void __populateSubCommandsFromJSON(SubCmdsCollection &subCmds, const std::string& jsonPath, const std::string& exeName)
+{
+    // parse JSON and add valid Sub Commands
+    pt::ptree jtree;
+    try {
+        pt::read_json(jsonPath,jtree);
+        std::string jsonSectionName;
+
+        if(exeName.compare("xbutil") == 0)
+            jsonSectionName = "xbutil_sub_commands";
+        else
```

Comment:
What happens if in the future if there is another executable that this code is used in (e.g., xbflash).  If it should occur then it would take the developer sometime to isolate and understand why their code doesn't work as expect.  My I suggest the following change:
```
if(exeName.compare("xbutil") == 0)
  jsonSectionName = "xbutil_sub_commands";

if(exeName.compare("xbmgmt") == 0)
  jsonSectionName = "xbmgmt_sub_commands";

if (jsonSectionName.empty())
  throw std::runtime_error("Error: Unknown executable '" + exeName + "'");
```

Doing gives the next developer a quick notification that there is something wrong.